zetta_utils/geometry/bbox.py

```diff
@@ -518,37 +567,67 @@ def get_size(self) -> Union[int, float]:  # pragma: no cover
 
     def aligned(self, other: BBox3D) -> tuple[bool, ...]:
         assert self.unit == other.unit
-        """Returns whether two BoundingBoxNDs are aligned, in
+        """Returns whether two BBox3Ds are aligned, in
         x_start, x_stop, y_start, y_stop, z_start, z_stop order."""
         return tuple(s[i] == o[i] for s, o in zip(self.bounds, other.bounds) for i in range(2))
 
     def contained_in(self: BBox3D, other: BBox3D) -> bool:
         assert self.unit == other.unit
-        """Returns whether the other BoundingBoxND contains self."""
+        """Returns whether the other BBox3D contains this one."""
         return all(
             (self_b[0] >= other_b[0] and other_b[1] >= self_b[1])
             for self_b, other_b in zip(self.bounds, other.bounds)
         )
 
     def intersects(self: BBox3D, other: BBox3D) -> bool:
+        """Returns whether the other BBox3D intersects this one.
+        Note that the `unit` property must match."""
         assert self.unit == other.unit
-        """Returns whether two BoundingBoxNDs intersect."""
         return all(
             (self_b[1] > other_b[0] and other_b[1] > self_b[0])
             for self_b, other_b in zip(self.bounds, other.bounds)
         )
 
     def intersection(self: BBox3D, other: BBox3D) -> BBox3D:
+        """Returns the intersection of another BBox3D with this one.
+        The `unit` property must match, but pprint_px_resolution is ignored.
+        The resulting bounds will be all `0` if they do not intersect.
+        """
         assert self.unit == other.unit
         if not self.intersects(other):
             return BBox3D(((0, 0), (0, 0), (0, 0)), unit=self.unit)
-        """Returns the intersection of two BoundingBoxNDs."""
         bounds = cast(
             tuple[Tuple2D, Tuple2D, Tuple2D],
             tuple((max(s[0], o[0]), min(s[1], o[1])) for s, o in zip(self.bounds, other.bounds)),
         )
         return BBox3D(bounds=bounds, unit=self.unit)
 
+    def union(self: BBox3D, other: BBox3D) -> BBox3D:
+        """Returns the union of another BBox3D with this one.
+        The `unit` property must match, but pprint_px_resolution is ignored.
+        """
+        assert self.unit == other.unit
```

Comment:
If it really is the union (rather than the supremum), then there should be `assert self.is_aligned(other)`.

Reply:
This comment really made me think!  I've always seen this operation (finding the smallest bounding box that contains both of the operand bounding boxes) called "union," but you're right, technically it's not a union in the sense of a non-box-shaped union of two arbitrary regions.  The area of the result here could be dramatically more than the sum of the areas of the inputs.

So I did some searching, but I can't find any references that don't call this operation on axis-aligned bounding boxes either "union" or (more rarely) "merge."  And I can't find any that apply the term "supremum" to bounding boxes at all — that generally means just the max of a set, which in this case I guess would be the point that's an upper bound for both bounding boxes.  I guess that's one corner of the result, with the other set to the infimum, but it still doesn't seem like a good term for what we're doing here.

I even went so far as to [ask ChatGPT](https://chatgpt.com/share/62d6e832-f218-40a3-9228-05d17c37b1a3) about it, figuring it has a pretty good grip on typical use of language.  It agrees this is a union operation (though also suggested merge).

So... how about we either leave this alone (as I think "union" is more standard), or call it `merge`, and in either case we make clear in the docstring that this is finding the minimum BBox3D that encloses both boxes?

Reply:
I still think it should be changed to `supremum` - it might also be good to make it a function that takes an iterable rather than a method.

So "supremum" generally refers to the least upper bound of a partially ordered (binary relation that's transitive and reflexive, not necessarily total) set, rather than the maximum. If you consider the sets in a topological space partially ordered under inclusion, then the supremum of two sets would be the union; for a set of numbers partially (and totally) ordered under the usual order, the supremum is the max of a set; for rectilinear bounding boxes partially ordered under inclusion, the supremum is the result of this operation. (Category theoretically, this is also called the "pushout" of two objects, which is dual to the "pullback" that one might encounter in other contexts.)

ChatGPT assumes that "supremum" given without context is in the category of sets of real numbers, since it probably has a lot more material on introductory real analysis than set theory, but you can get it to print the correct answer: 
![Screenshot from 2024-08-22 15-15-00](https://github.com/user-attachments/assets/bbcc8997-d0d6-4de7-9bb8-c0ea77731aef)

Reply:
Okey dokey; renamed to supremum.  I left it as a method to match the interface of `intersection`, and mentioned "union" in the docstring in case somebody goes looking for it with that term.

Reply:
I'll approve - though mypy is failing at the moment with the name change.

Reply:
I can tell you that in the game industry, at least, this operation is very common and is universally called union (whether it should be or not).

But I don't want to argue about it; as long as the word is in there, it'll be found when somebody searches for it, and if they're confused by this comment, they can talk to you about it. :)  Wording updated.